eval: save_tv_as_string: Correctly handle an empty string

jamessan opened the pull request:
When tv_get_string_chk returns a non-NULL value, we have a valid string.
Propagating an error state (*len = -1, NULL return) for an empty string
is invalid.

Closes #6554

jamessan commented:
@ZyX-I Any idea why the test is sometimes failing in Travis like this:

```
[ RUN      ] system() passing input works with an empty string: ERR
./test/functional/helpers.lua:93: Failed to evaluate expression
stack traceback:
	./test/functional/helpers.lua:93: in function 'eval'
	...build/neovim/neovim/test/functional/eval/system_spec.lua:258: in function <...build/neovim/neovim/test/functional/eval/system_spec.lua:257>
```

I've been running `system_spec.lua` in a loop on my VM here and haven't had any issues so far.  In my previous version of the PR, I tested cherry-picking the commit onto code before the big typval refactor and it seemed to work fine there.

jamessan commented:
Hmm, after clearing Travis' cache, everything ran fine except for the flaky timers tests.

justinmk commented:
Let's re-run it a few times...

jamessan commented:
Ok, it happened again in the coverage build.

There was also this stderr output reported at the end of the run:

```
-- Output to stderr:
2017/04/28 16:34:19 ERROR 10786/open_log_file:127: Couldn't open USR_LOG_FILE, logging to stderr! This may be caused by attempting to LOG() before initialization functions are called (e.g. init_homedir()).
2017/04/28 16:34:19 ERROR 10786/call_set_error:723: RPC: ch 1 was closed by the client
2017/04/28 16:34:41 ERROR 11266/open_log_file:127: Couldn't open USR_LOG_FILE, logging to stderr! This may be caused by attempting to LOG() before initialization functions are called (e.g. init_homedir()).
2017/04/28 16:34:41 ERROR 11266/call_set_error:723: RPC: ch 1 was closed by the client
```

justinmk commented on src/nvim/eval.c:
```diff
@@ -17409,7 +17409,8 @@ static char *save_tv_as_string(typval_T *tv, ptrdiff_t *const len, bool endnl)
     const char *ret = tv_get_string_chk(tv);
-    if (ret && (*len = strlen(ret))) {
+    if (ret) {
+      *len = strlen(ret);
```
Might be worth temporarily reverting this change to see if the problem still occurs with the new test. To avoid having to look at a bunch of failed builds, change the expect result to what happens on master.

```lua
eq("", eval('system("echo test", "")'))
```

jamessan replied:
No lua tracebacks yet with the code reverted...

justinmk commented on test/functional/eval/system_spec.lua:
```diff
@@ -255,7 +255,7 @@ describe('system()', function()
       eq(2, eval("1+1"))  -- Still alive?
```
off topic, just noticed the test on lin 250 should be:

    feed_command('call system("cat - &", "input")')

jamessan committed:
eval: save_tv_as_string: Correctly handle an empty string

When tv_get_string_chk returns a non-NULL value, we have a valid string.
Propagating an error state (*len = -1, NULL return) for an empty string
is invalid.

Closes #6554

jamessan commented:
I've run the Travis builds a few times now and none of them have failed.  Looking better now.

justinmk commented:
Hmm, saw this in https://github.com/neovim/neovim/pull/7924 :

```
system() passing input works with an empty string: [1m[31mERR[0m[0m
test/functional/helpers.lua:93: Failed to evaluate expression

stack traceback:
	test/functional/helpers.lua:93: in function 'eval'
	...build/neovim/neovim/test/functional/eval/system_spec.lua:262: in function <...build/neovim/neovim/test/functional/eval/system_spec.lua:261>
```

justinmk commented:
I can repro it locally, will investigate.
